content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,110 @@
+2. Create a new RNG and pass it around using the [`np.random.default_rng`](https://numpy.org/doc/1.18/reference/random/generator.html#numpy.random.default_rng) function.
+3. Be careful with parallel computations and rely on [numpy strategies for reproducible parallel number generation](https://numpy.org/doc/1.18/reference/random/parallel.html).
+
+Note that with numpy <1.17 the way to create a new RNG is to use [`np.random.RandomState`](https://numpy.org/doc/1.18/reference/random/legacy.html#numpy.random.RandomState) which is based on the popular Mersenne Twister 19937 algorithm. This is also how the global numpy RNG is created. It is still possible to use this function in versions higher than 1.17 but it is now recommended to use `default_rng` which returns an instance of the statistically better [PCG64](https://www.pcg-random.org) RNG.
+
+## Random number generation with numpy
+When you import `numpy` in your python script a RNG is created behind the scenes. This RNG is the one used when you generate a new random value using a function such as `np.random.random`. I will here refer to this RNG as the global numpy RNG.
+
+Although not recommended, it is a common practice to reset the seed of this global RNG at the beginning of a script using the `np.random.seed` function. Fixing the seed at the beginning ensures that the script is reproducible: the same values and results will be produced each time you run it. However, although sometimes convenient, using the global numpy RNG is considered a bad practice. A simple reason is that using global variables can lead to undesired side effects. For instance one might use `np.random.random` without knowing that the seed of the global RNG was set somewhere else in the codebase. Quoting the [Numpy Enhancement Proposal (NEP) 19](https://numpy.org/neps/nep-0019-rng-policy.html) by Robert Kern about the numpy RNG policy:
+
+> The implicit global RandomState behind the `np.random.*` convenience functions can cause problems, especially when threads or other forms of concurrency are involved. Global state is always problematic. We categorically recommend avoiding using the convenience functions when reproducibility is involved. [...] The preferred best practice for getting reproducible pseudorandom numbers is to instantiate a generator object with a seed and pass it around.
+
+In short:
+* Instead of using `np.random.seed`, which reseeds the already created global numpy RNG and then using `np.random.*` functions you should create a new RNG.
+* You should create one RNG at the beginning of your script (with a seed if you want reproducibility) and use this RNG in the rest of your script.
+
+To create a new RNG you can use the `default_rng` function as illustrated in the [introduction of the random module documentation](https://numpy.org/doc/1.18/reference/random/index.html#introduction):
+
+```python
+import numpy as np
+seed = 12345
```

Comment:
Can you change that (and bellow) to a more random value? This is to prevent people that copy paste to 1. copy all the time the same value (hence arguably not have a RNG anymore and have the same number generator as everyone), 2. the doc suggest that larger numbers should be used to have a better entropy (though there are better safeguard now as opposed to the old infra as I understand it.)

Reply:
> though there are better safeguard now as opposed to the old infra as I understand it.

yes I thought that with the new seed sequence algorithm we were sure to get a good initial state for the RNG, even with a bad input seed from the user?

Reply:
> Can you change that (and bellow) to a more random value?

sorry can you elaborate a bit? you mean something that looks a bit more random like `74329087451`?

Reply:
Ok I found more information about that at https://numpy.org/doc/stable/reference/random/index.html#random-sampling-numpy-random